### PR TITLE
Reject receipt requests with empty room or event IDs

### PR DESCRIPTION
**Pull request by Fizzadar**
Currently synapse accepts and stores these in the `receipts_linearized` table - the insertion is not validated in the transaction I believe because the server may not yet know of the event ID. Rejecting empty values here matches the spec.

Signed off by Nick @ Beeper (@fizzadar).

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog).
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/rest/client/receipts.py [hunk 1]
```diff
@@ -56,8 +56,8 @@ async def on_POST(
     ) -> Tuple[int, JsonDict]:
         requester = await self.auth.get_user_by_req(request)
 
-        if not room_id or not event_id:
-            raise SynapseError(400, "A room ID and event ID must be specified")
+        if not RoomID.is_valid(room_id) or not event_id.startswith(EventID.SIGIL):
```
Comment (Fizzadar) [inline synapse/rest/client/receipts.py:59]: Not super happy with the `startswith` but `EventID.is_valid` is broken for non-V1/2 event IDs so unsure how to proceed.

**Comment by Fizzadar** [comment]
> This is a nice improvement! We could probably check that the room / event IDs are of the proper form at least?

Definitely, added https://github.com/matrix-org/synapse/pull/14632/commits/b5235361a0e330d66c2bc38e8acbe091ccaead61 (& associated GH comment).

> > the insertion is not validated in the transaction I believe because the server may not yet know of the event ID.
> 
> Since this is the client-server API the room and event ID _must_ be known to the server since the client already got them from the server? (Modulo room retention stuff.)

Hm yeah that's a good point, looks like this was changed in https://github.com/matrix-org/synapse/commit/9f797a24a452a513628263b1b03172cee20a9856 although I'm not sure on why that might happen, pretty old commit...

**Comment by clokep** [comment]
> Hm yeah that's a good point, looks like this was changed in [9f797a2](https://github.com/matrix-org/synapse/commit/9f797a24a452a513628263b1b03172cee20a9856) although I'm not sure on why that might happen, pretty old commit...

Oh good find. I suspect the reasoning for that is lost in the depths of time. 😢

File: synapse/rest/client/receipts.py (hunk 1 above)
Comment (clokep) [reply synapse/rest/client/receipts.py:59]: Filed #14638.

File: tests/rest/client/test_receipts.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
+# Copyright 2018-2019 New Vector Ltd
+# Copyright 2019 The Matrix.org Foundation C.I.C.
```
Comment (clokep) [inline tests/rest/client/test_receipts.py:2]:
```suggestion
# Copyright 2022 The Matrix.org Foundation C.I.C.
```

File: changelog.d/14632.misc [hunk 3]
```diff
@@ -0,0 +1 @@
+Reject invalid read receipt requests with empty room or event IDs. Contributed by Nick @ Beeper (@fizzadar).
```
Comment (clokep) [inline changelog.d/14632.misc:1]: I'm half tempted to say this is a bugfix, but 🤷

**Review by clokep — COMMENTED** [review]
Looks good, minus the nits. I can't seem to apply them myself though!

File: changelog.d/14632.misc (hunk 3 above)
Comment (Fizzadar) [reply changelog.d/14632.misc]: Ah good spot, agreed! https://github.com/matrix-org/synapse/pull/14632/commits/fbb7b16148fcc56ac55b6ccd53bcc857987611cd

**Comment by Fizzadar** [comment]
Huh, GH isn't showing the option to enable maintainer edits either 🤔 might be because it's from our org fork? Anyway, applied the nit!

**Review by clokep — APPROVED** [review]

**Comment by clokep** [comment]
Looks grand, thanks! 👍 Might need to enable that at an org level or something? Thanks for applying everything quickly!

**Pull request merged**